src/Hazelcast.Net/Serialization/Compact/CompactOptions.cs

```diff
@@ -0,0 +1,462 @@
+
+            EnsureUniqueTypeNamePerSerializedType(serializedType, typeName);
+            EnsureUniqueSchemaPerTypeName(schema);
+            EnsureTypeNameSerializerCanSerializeType(typeName, serializedType);
+
+            _serializedType.Add(serializedType);
+            _serializedType_typeName[serializedType] = typeName;
+            _typeName_schema[schema.TypeName] = schema;
+
+            _typeName_isClusterSchema[schema.TypeName] = isClusterSchema;
+            _serializedType_isClusterSchema[serializedType] = isClusterSchema;
+        }
+
+        /// <summary>
+        /// Configures a schema for a serialized type.
+        /// </summary>
+        /// <param name="serializedType">The serialized type.</param>
+        /// <param name="isClusterSchema">Whether the schema is known by the cluster already.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="serializedType"/> is <c>null</c>.</exception>
+        internal void SetSchema(Type serializedType, bool isClusterSchema)
+        {
+            if (serializedType == null) throw new ArgumentNullException(nameof(serializedType));
+
+            _serializedType.Add(serializedType);
+            _serializedType_isClusterSchema[serializedType] = isClusterSchema;
+        }
+
+        /// <summary>
+        /// Configures a schema.
+        /// </summary>
+        /// <param name="schema">The schema.</param>
+        /// <param name="isClusterSchema">Whether the schema is known by the cluster already.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="schema"/> is <c>null</c>.</exception>
+        /// <exception cref="ConfigurationException">The operation conflicts with information that were already provided.</exception>
+        internal void SetSchema(Schema schema, bool isClusterSchema)
+        {
+            if (schema == null) throw new ArgumentNullException(nameof(schema));
+
+            EnsureUniqueSchemaPerTypeName(schema);
+
+            _typeName_schema[schema.TypeName] = schema;
+            _typeName_isClusterSchema[schema.TypeName] = isClusterSchema;
+        }
+
+        private bool GetIsClusterSchema(Type serializedType, string typeName)
+        {
+            var isClusterSchema = false;
+            if (_serializedType_isClusterSchema.TryGetValue(serializedType, out var ics1))
+                isClusterSchema |= ics1;
+            if (_typeName_isClusterSchema.TryGetValue(typeName, out var ics2))
+                isClusterSchema |= ics2;
+            return isClusterSchema;
+        }
+
+        /// <summary>
+        /// Creates and returns the serializer registrations corresponding to options.
+        /// </summary>
+        /// <returns>The serializer registrations corresponding to options.</returns>
+        /// <exception cref="ConfigurationException">The options contain conflicts that prevent the registrations from being created.</exception>
+        internal IEnumerable<CompactRegistration> GetRegistrations()
```

Comment:
Just wondering, probably answer is in other PRs. There are other type serializers between compact and no-config compact. However, `GetRegistrations` handles them all together.  I think there will be another layer between serialization and registration.

Reply:
`GetRegistration` in invoked when initializing the compact serialization service - then when we lookup for a serializer, we ask the compact serialization service whether it has a registration for the object's type. if not, we move on and try custom serializers, portable, etc. If, in the end, we still don't have a serializer, we'll ask the compact serialization service to please create a registration for that type and handle it from now on.